packages/@aws-containers/aws-ecs-builder/lib/extensions/appmesh.ts

```diff
@@ -0,0 +1,216 @@
+import * as appmesh from '@aws-cdk/aws-appmesh';
+import * as ec2 from '@aws-cdk/aws-ec2';
+import * as ecr from '@aws-cdk/aws-ecr';
+import * as ecs from '@aws-cdk/aws-ecs';
+import * as cdk from '@aws-cdk/core';
+import { Service } from '../service';
+import { Container } from './container';
+import { ServiceExtension, ServiceBuild } from './extension-interfaces';
+
+/**
+ * The settings for the App Mesh extension
+ */
+export interface MeshProps {
+  /**
+   * The service mesh into which to register the service
+   */
+  readonly mesh: appmesh.Mesh;
+}
+
+/**
+ * This extension adds an Envoy sidecar to the task definition and
+ * creates the App Mesh resources required to route network traffic
+ * to the container in a service mesh
+ */
+export class AppMeshExtension extends ServiceExtension {
+  protected virtualNode!: appmesh.VirtualNode;
+  protected virtualService!: appmesh.VirtualService;
+  private mesh: appmesh.Mesh;
+
+  constructor(props: MeshProps) {
+    super('appmesh');
+    this.mesh = props.mesh;
+  }
+
+  public prehook(service: Service, scope: cdk.Construct) {
+    this.parentService = service;
+    this.scope = scope;
+
+    // Make sure that the parent cluster for this service has
+    // a namespace attached.
+    if (!this.parentService.cluster.defaultCloudMapNamespace) {
+      this.parentService.cluster.addDefaultCloudMapNamespace({
+        name: 'internal',
+      });
+    }
+  }
+
```

Comment:
Should namespace be it's own object that is a child of an environment (and within which services are placed)? I don't think folks necessarily equate a namespace to a cluster. 

Conventionally I like that a namespace with an easy name like `internal` is a nice default. The catch I can think of is the name is unique within an account and region. Ideally you'd be able to have two separate environments, but both use the `.internal` moniker, without any conflict between them. Let me bring this up to the Cloud Map team.

Reply:
I added an integration test and verified that this did in fact error at the Cloud Map level on when multiple environments were created on the same account. To fix it I changed the code to use a namespace named after the environment. (There is already protection at the CDK level against creating multiple environments with the same name)